spec/p2p: document the p2p API used by Reactors (#851)
cometbft/cometbft: spec/p2p/reactor/p2p-api.md

It manages all the reactors running in a node, registered during the node setup phase
(see the reactor's [Registration section][reactor-registration]),
and keeps track of the connections with peers.

## `Switch` API

The table below summarizes the interaction of the standard reactors with the
p2p layer, by listing the [`Switch`][switch-type] methods used by them:

| `Switch` API method                                     | consensus | block sync | state sync | mempool | evidence  | PEX   |
|--------------------------------------------|-----------|------------|------------|---------|-----------|-------|
| `Peers() IPeerSet`                         | x         | x          |            |         |           | x     |
| `NumPeers() (int, int, int)`               |           | x          |            |         |           | x     |
| `Broadcast(Envelope) chan bool`            | x         | x          | x          |         |           |       |
| `MarkPeerAsGood(Peer)`                     | x         |            |            |         |           |       |
| `StopPeerForError(Peer, interface{})`      | x         | x          | x          | x       | x         | x     |

> | `StopPeerGracefully(Peer)`                 |           |            |            |         |           | x     |

Reviewer: These functions (MarkPeerAsGood, StopPeer*) seem to imply non-trivial behavior (that also involve other reactors). This might be interesting to formalize (in quint).

Also, I understand that this is how it is right now. But would it be thinkable that these three functions move from the Switch API to the Peer API below? (I just try to understand the wiggle room)

Reviewer: Connected to my above comment: I would like to understand whether in the future we would need the Switch API at all:
- maintaining the responsibility to maintain the information provided by Peers and NumPeers could be moved to the reactors
- broadcast could be implemented within the reactors using send
- Reactor may be replaced by a different mechanism to switch between protocols (syncing protocols to consensus).

Author: This discussion is being addressed here: https://github.com/cometbft/knowledge-base/pull/8

Author: The document referring to the separation of `Switch` and `Peer` is not a WIP, please refer to: https://github.com/cometbft/knowledge-base/blob/51eb728dae2f5535c902275297e4aa1e6a600531/p2p/reactors/switch-peer.md

Consensus reactor implementation, from which it uses the exported
`SwitchToConsensus()` method.
While available, this inter-reactor interaction approach is discouraged and
should be avoided, as it violates the assumption that reactors are independent.

## `Peer` API

The table below summarizes the interaction of the standard reactors with
connected peers, by listing the [`Peer`][peer-interface] interface methods used
by them:

| `Peer` API method                                     | consensus | block sync | state sync | mempool | evidence  | PEX   |
|--------------------------------------------|-----------|------------|------------|---------|-----------|-------|
| `ID() ID`                                  | x         | x          | x          | x       | x         | x     |
| `IsRunning() bool`                         | x         |            |            | x       | x         |       |
| `Quit() <-chan struct{}`                   |           |            |            | x       | x         |       |
| `Get(string) interface{}`                  | x         |            |            | x       | x         |       |

> | `Set(string, interface{})`                 | x         |            |            |         |           |       |

Author: This is a relevant discussion, being addressed here: https://github.com/cometbft/knowledge-base/pull/8

# API for Reactors

This document describes the API provided by the p2p layer to the protocol
layer, namely to the registered reactors.

This API consists of two interfaces: the one provided by the `Switch` instance,
and the ones provides by multiple `Peer` instances, one per connected peer.
The `Switch` instance is provided to every reactor as part of the reactor's
[registration procedure][reactor-registration].
The multiple `Peer` instances are provided to every registered reactor whenever

> a [new connection with a peer][reactor-addpeer] is established.

Reviewer 2: Can we include this explanation in the text?

implementation.  It manages all the reactors running in a node and keeps track
of the connections with peers.
The table below summarizes the interaction of the standard reactors with the `Switch`:

| `Switch` API method                                     | consensus | block sync | state sync | mempool | evidence  | PEX   |
|--------------------------------------------|-----------|------------|------------|---------|-----------|-------|
| `Peers() IPeerSet`                         | x         | x          |            |         |           | x     |
| `NumPeers() (int, int, int)`               |           | x          |            |         |           | x     |
| `Broadcast(Envelope) chan bool`            | x         | x          | x          |         |           |       |
| `MarkPeerAsGood(Peer)`                     | x         |            |            |         |           |       |
| `StopPeerForError(Peer, interface{})`      | x         | x          | x          | x       | x         | x     |
| `StopPeerGracefully(Peer)`                 |           |            |            |         |           | x     |

> | `Reactor(string) Reactor`                  |           | x          |            |         |           |       |

Reviewer 2: Similar to @josef-widder 's comment above, do we want to keep this method moving forward? (it seems to enable implicit assumptions on the reactors registered, thus hindering modularity)

Reviewer 2: I see you're explaining further below that its usage is discouraged. I think an output of all these spec-related PRs could a list of things we don't like, which can then be turned into a tracking issue that we can prioritize and plan as tech-debt.

Reviewer: It is basically the idea to collect all the points we don't like or need to discuss whether we like them in https://github.com/cometbft/knowledge-base/pull/8.

The above list is not exhaustive as it does not include all the `Switch` methods
invoked by the PEX reactor, a special component that should be considered part
of the p2p layer. This document does not cover the operation of the PEX reactor
as a connection manager.

### Peers State

The first two methods in the switch API allow reactors to query the state of
the p2p layer: the set of connected peers.

> 
>     func (sw *Switch) Peers() IPeerSet
> 

Reviewer 3: ```suggestion
   ```go
    func (sw *Switch) Peers() IPeerSet
   ```
```

Author: Is the rendering odd here? It looks the same as in the other uses of this code block.

Author: The point is that I would need to change this thorough the document, maybe also in the other file.

Do you think that having specific Go-code highlighting renders it easier to read?

Reviewer 3: I do, but not too much for such small blocks.
Update only if you get free cycles.

The `Peers()` method returns the current set of connected peers.
The returned `IPeerSet` is an immutable concurrency-safe copy of this set.
Observe that the `Peer` handlers returned by this method were previously
[added to the reactor][reactor-addpeer] via the `InitPeer(Peer)` method,
but not yet removed via the `RemovePeer(Peer)` method.
Thus, a priori, reactors should already have this information.

    func (sw *Switch) NumPeers() (outbound, inbound, dialing int)

The `NumPeers()` method returns the current number of connected peers,

> distinguished between `outbound` and `inbound` peers.

Author: This is a good point.

> distinguished between `outbound` and `inbound` peers.

Author: Added a comment regarding this API method.

An `outbound` peer is a peer the node has dialed to, while an `inbound` peer is
a peer the node has accepted a connection from.
The third field `dialing` reports the number of peers to which the node is

> currently attempting to connect, so not (yet) connected peers.

Reviewer 2: How should this be relevant for a reactor?

Author: The PEX reactor uses this field, the others don't. The PEX reactor is a really different component, that in the current version is literally part of the p2p layer.

Reviewer: We should really try to find a place for the PEX reactor. In some sense it is a reactor, in some sense it is part of p2p. I think this kind of questions all come from the way PEX is situated.

Reviewer 2: I see. This is one of the first things we will need to fix in the p2p layer when we tackle it. My high-level understanding is that the PEX "reactor" should be using a different interface than the rest, so as not to "pollute" the reactor interface with things that only make sense for PEX.

### Broadcast

The switch provides, mostly for historical or retro-compatibility reasons,
a method for sending a message to all connected peers:

    func (sw *Switch) Broadcast(e Envelope) chan bool

The `Broadcast()` method is not blocking and returns a channel of booleans.
For every connected `Peer`, it starts a background thread for sending the
message to that peer, using the `Peer.Send()` method
(which is blocking, as detailed in [Send Methods](#send-methods)).
The result of each unicast send operation (success or failure) is added to the

> returned channel, which is closed when all operations are completed.

Reviewer 3: is there a way to connect the result with the peer?

Author: No. Essentially, the only point is to know how many sends have failed.

Reviewer 2: Do reactors use this info?

Author: No, all the calls in Consensus, State Sync, and Block Sync reactors ignore the return value.

Reviewer 2: So, candidate for removal :-)

Reviewer 2: So, candidate for removal :-)
(not in this PR)

Author: Highlighted this information (no one is using this returning value) in the document.

> _Note_: while this method is helpful, being adopted by several standard
> reactors, its current _implementation_ is not efficient, as the marshalling
> of the provided message is performed as part of the `Peer.Send()` helper
> method, that is, once per connected peer.

### Vetting Peers

The p2p layer relies on the registered reactors to gauge the _quality_ of peers.
The following method can be invoked by a reactor to inform the p2p layer that a
peer has presented a "good" behaviour.
This information is registered in the node's address book and influences the
operation of the Peer Exchange (PEX) protocol, as node discovery adopts a bias
towards "good" peers:

    func (sw *Switch) MarkPeerAsGood(peer Peer)

At the moment, it is up to the consensus reactor to vet a peer.

> A peer is marked as good when the consensus protocol receives

Author: ```suggestion
A peer is marked as good whenever the consensus protocol collects a multiple of
```

`votesToContributeToBecomeGoodPeer = 10000` useful votes or

> `blocksToContributeToBecomeGoodPeer = 10000` useful block parts from that peer.

Reviewer 3: Does the counter decay with time, or is it a matter of time for any non byzantine node to be marked as good?

Reviewer 3: I don't mean the values of `votesToContributeToBecomeGoodPeer` and `blocksToContributeToBecomeGoodPeer` but the counters used to track if the peers have reached these values. But the fact that 
>  The number of peers with this state is limited.
would have a somewhat similar effect.

Author: Link to the knowledge-base repository added, I think this discussion should continue there.

duplicated or redundant messages, fairly common in gossip-based communication,
are not counted as "useful".

### Stopping Peers

Reactors can instruct the p2p layer to disconnect from a peer.
Using the p2p layer's nomenclature, the reactor requests a peer to be stopped.
The peer's send and receive routines are in fact stopped, interrupting the
communication with the peer.
The `Peer` is then [removed from every registered reactor][reactor-removepeer],
using the `RemovePeer(Peer)` method, and from the set of connected peers.

    func (sw *Switch) StopPeerForError(peer Peer, reason interface{})

All the standard reactors employ the above method for disconnecting from a peer
in case of errors.
These are errors that occur when processing a message received from a `Peer`.
The produced `error` is provided to the method as the `reason`.

The `StopPeerForError()` method has an important *caveat*: if the peer to be
stopped is configured as a _persistent peer_, the switch will attempt
reconnecting to that same peer.
While this behaviour makes sense when the method is invoked by other components
of the p2p layer (e.g., in the case of communication errors), it does not make

> sense when it is invoked by a reactor.

Reviewer: I guess reconnection might also happen as a result of PEX querying seed nodes for peer addresses, that is, the peer is not put on a "blocked list" for the future, but just removed at this moment. If this is true, perhaps we can add a sentence about that here.

|--------------------------------------------|-----------|------------|------------|---------|-----------|-------|
| `ID() ID`                                  | x         | x          | x          | x       | x         | x     |
| `IsRunning() bool`                         | x         |            |            | x       | x         |       |
| `Quit() <-chan struct{}`                   |           |            |            | x       | x         |       |
| `Get(string) interface{}`                  | x         |            |            | x       | x         |       |
| `Set(string, interface{})`                 | x         |            |            |         |           |       |
| `Send(Envelope) bool`                      | x         | x          | x          | x       | x         | x     |
| `TrySend(Envelope) bool`                   | x         | x          |            |         |           |       |

The above list is not exhaustive as it does not include all the `Peer` methods
invoked by the PEX reactor, a special component that should be considered part
of the p2p layer. This document does not cover the operation of the PEX reactor

> as a connection manager.

Reviewer: Does it then make sense to have the PEX column here?

sharing peer-specific state between reactors:

    func (p Peer) Get(key string) interface{}
    func (p Peer) Set(key string, data interface{})

This key-value store can be seen as an asynchronous mechanism to exchange the
state of a peer between reactors.
In the current use-case of this mechanism, the Consensus reactor populates the
key-value store with a `PeerState` instance for each connected peer.
The Consensus reactor routines interacting with a peer read and update the
shared peer state.
The Evidence and Mempool reactors, in their turn, periodically query the
key-value store of each peer for retrieving, in particular, the last height
reported by the peer.
This information, produced by the Consensus reactor, influences the interaction
of these two reactors with their peers.

### Send methods

Finally, a `Peer` instance allows a reactor to send messages to companion
reactors running at that peer.
This is ultimately the goal of the switch when it provides `Peer` instances to
the registered reactors.
There are two methods for sending messages:

    func (p Peer) Send(e Envelope) bool
    func (p Peer) TrySend(e Envelope) bool

> The two message-sending methods receive an `Envelope`, which content should be

Reviewer 2: ```suggestion
The two message-sending methods receive an `Envelope`, whose content should be
```
